README.md
# LocalCache

[![Build Status](https://secure.travis-ci.org/hi-man/localcache.png)](https://travis-ci.org/hi-man/localcache)

now, let's back to our classroom,
professor is talking about the architecture of `cpu`,
just like the software your are creating，`Redis` is the `L2` cache,
and `Yac` is the `L1` cache for your software,
and `LocalCache` is the cache manager of `Redis` and `Yac`.

so, if your redis server is busying, and your data does not change frequently,
`LocalCache` is the right man you are finding.

> If I have seen further, it is by standing on the shoulders of giants.

thanks to [phpredis](https://github.com/phpredis/phpredis) and [yac](https://github.com/laruence/yac)

<!-- vim-markdown-toc GFM -->

- [Feature](#feature)
- [Usage](#usage)
- [Requirement](#requirement)
- [Install](#install)
- [testing](#testing)
- [Classes](#classes)
  - [CachePoolService](#cachepoolservice)
    - [initCacheInstance](#initcacheinstance)
    - [getCacheInstance](#getcacheinstance)
    - [setCacheValue](#setcachevalue)
    - [getCacheValue](#getcachevalue)
    - [deleteByKey](#deletebykey)
  - [CacheService](#cacheservice)
    - [getConfigByConnection](#getconfigbyconnection)
    - [logException](#logexception)
    - [setCacheValue](#setcachevalue-1)
    - [getCacheValue](#getcachevalue-1)
    - [deleteByKey](#deletebykey-1)
  - [LocalCache](#localcache)
    - [Construct](#construct)
    - [select](#select)
    - [get](#get)
    - [set](#set)
    - [delete / unlink](#delete--unlink)
    - [expire](#expire)
    - [clear](#clear)
    - [hDel / hGetAll / hMSet / hMGet / hSet / hGet](#hdel--hgetall--hmset--hmget--hset--hget)
    - [setLocalCacheTimeout](#setlocalcachetimeout)
    - [getLocalCacheTimeout](#getlocalcachetimeout)

<!-- vim-markdown-toc -->

# Feature

- [x] a redis client with retry mechanism
- [x] reduce redis requests by caching data in memory
- [x] support redis connection pool
- [x] more graceful exception handler
- [x] disable cache on demands

# Usage

1. extend class `CacheService` (recommend)
2. use static funcitons in `CachePoolService`
3. use `LocalCache` class directly

# Requirement

- PHP 7.0+
- phpredis
- yac

# Install

```
composer require 'hi-man/localcache'
```

# testing

```
composer test
```

# Classes

## CachePoolService

---

pool of `LocalCache`, can be called statically

### initCacheInstance

initialize `LocalCache`

### getCacheInstance

get instance of `LocalCache` by connection name

### setCacheValue

set cache value

### getCacheValue

get cache value

### deleteByKey

delete from cache by key

## CacheService

---

abstract class to use CachePoolService more productive

### getConfigByConnection

**MUST implemented** get redis configuration by connection identifier

### logException

**MUST implemented** exception handler

### setCacheValue

wrapper of `CachePoolService::setCacheValue`

### getCacheValue

wrapper of `CachePoolService::getCacheValue`

### deleteByKey

wrapper of `CachePoolService::deleteByKey`

## LocalCache

---

provide a local cache between application and redis server

### Construct

```php
$lc = new LocalCache(
  '127.0.01' /* redis host */,
  'Yac prefix' /* yac prefix, empty prefix will disable yac, default value is empty, max length is 20 */,
  6379 /* redis port, default value is 6379 */,
  3 /* redis connection timeout, in seconds, default value is 3 */,
  500000 /* redis retry interval, in microseconds, default value is 500000 */,
  3 /* redis read timeout, default value is 3 */,
  3 /* max retry, default value is 3 */,
  0 /* redis reserved, default value is 0 */
);
```

### select

the same as redis command `select`, but not really issue a command request.

```php
$lc->select(0 /* redis database index */);
```

### get

the same as redis command `get`, use yac cache value first, then issue a command request if cache is missing.

```php
$lc->get(
  'key' /* redis item key */,
  'default value' /* default value if the key does not exists */
);
```

### set

the same as redis command `set`, reset yac cache value

```php
$lc->set(
    'key',      /* redis item key */
    'value',    /* value to store */
    3           /* ttl */
    'default value' /* default value if the key does not exists */
);

```

### delete / unlink

the same as redis command `delete` or `unlink`, also delete yac cache

```php
$lc->delete(
    'key',      /* redis item key */
);

$lc->unlink(
    'key',      /* redis item key */
);
```

### expire

the same as redis command `expire`, also reset yac cache expire time

```php
$lc->expire('key' /* redis item key */, 3 /* expire time in seconds */);
```

### clear

the same as redis command `flushdb`, but flush **all** yac cache

```php
$lc->clear();
```

### hDel / hGetAll / hMSet / hMGet / hSet / hGet

the same as redis command

### setLocalCacheTimeout

set yac cache timeout, **set the right value for your scenario**.
cache invalidation is a big concept to deal with.

```php
$lc->setLocalCacheTimeout(
    'key',      /* redis item key */
    'value',    /* value to store */
    3           /* ttl */
    'default value' /* default value if the key does not exists */
);

```

### getLocalCacheTimeout

get yac cache timeout

```php
$lc->getLocalCacheTimeout();
```


CHANGELOG.md
# Release Notes

## v0.0.9(2020-11-28)

- yacDeleteHash is a public function now

## v0.0.8(2020-03-27)

- use unlink instead of delete

## v0.0.7(2020-01-26)

- add cache key prefix for commands:
  hdel / hgetall / hmset / hmget / hset / hget / expire / get / set / delete
- ignore yac fails in SET command

## v0.0.6(2019-11-13)

- support hDel / hGetAll / hMSet / hMGet / hSet / hGet

## v0.0.5(2019-09-07)

- support key length greater than 48
- initialize default connection and database index

## v0.0.4(2019-09-03)

- support exception function
- support pool with or without yac

## ~v0.0.3 (2019-08-20)~

- add: support all redis commands
- fix: use the right yac key with prefix

## ~v0.0.2 (2019-08-15)~

- add: `CachePoolService` a cache pool identified by redis connection name
- add: `CacheService` a cache abstract class use cache pool more productive
- optimized: redis operation more effective

## v0.0.1 (2019-08-09)

- add: `LocalCache` class provide a local cache between application and redis server


tests/localcacheTest.php
<?php
namespace Tests;

use LocalCache\LocalCache;
use PHPUnit\Framework\TestCase;

final class LocalCacheTest extends TestCase
{
    private $localcache;

    protected function setUp()
    {
        $this->localcache = new LocalCache(
            'localhost',
            'yacPrefixTesting',
            6379,
            3,
            500000,
            3,
            3,
            0
        );

        $this->localcache->setLocalCacheTimeout(3);
    }

    public function testGetDefaultValue()
    {
        $value = 'default_value';
        $this->localcache->select(3);
        $this->localcache->clear();

        $ret = $this->localcache->get('key_doest_not_exists', $value);
        $this->assertTrue($ret === false);
    }

    public function testGetValueWithExpiretime()
    {
        $this->localcache->select(3);
        $this->localcache->clear();

        $key = 'getValueWithExpiretime';
        $value = 'getValueWithExpiretimeValue';
        $ttl = 2;

        $this->localcache->set($key, $value, $ttl);
        sleep($ttl + 1);
        $ret = $this->localcache->get($key);
        $this->assertTrue($ret === false);

        $this->localcache->set($key, $value, $ttl);
        $ret = $this->localcache->get($key);
        $this->assertTrue($ret === $value);
        sleep($ttl + 1);
        $ret = $this->localcache->get($key);
        $this->assertTrue($ret === false);
    }

    public function testGetValueWithoutExpiretime()
    {
        $this->localcache->select(3);
        $this->localcache->clear();

        $key = 'getValueWithoutExpiretime';
        $value = 'getValueWithoutExpiretimeValue';

        $this->localcache->set($key, $value);
        $ret = $this->localcache->get($key);
        $this->assertTrue($ret === $value);
    }

    public function testDeleteKey()
    {
        $this->localcache->select(3);
        $this->localcache->clear();

        $key = 'deletekey';
        $value = 'deletekeyvalue';

        $this->localcache->set($key, $value);
        $ret = $this->localcache->get($key);
        $this->assertTrue($ret === $value);
        $this->localcache->delete($key);
        $ret = $this->localcache->get($key);
        $this->assertTrue($ret === false);
    }

    public function testClear()
    {
        $this->localcache->select(3);
        $this->localcache->clear();

        $key = 'deletekey';
        $value = 'deletekeyvalue';

        $this->localcache->set($key, $value);
        $ret = $this->localcache->get($key);
        $this->assertTrue($ret === $value);
        $this->localcache->clear();
        $ret = $this->localcache->get($key);
        $this->assertTrue($ret === false);
    }

    public function testHashSingle()
    {
        $this->localcache->select(3);
        $this->localcache->clear();

        $key = 'hashkey';
        $fieldkey = 'fieldkey';
        $value = 'value';

        $ret = $this->localcache->hSet($key, $fieldkey, $value);
        $this->assertTrue($ret !== false);
        $ret = $this->localcache->hGet($key, $fieldkey);
        $this->assertTrue($ret === $value);

        sleep($this->localcache->getLocalCacheTimeout() + 1);
        $ret = $this->localcache->hGet($key, $fieldkey);
        $this->assertTrue($ret === $value);
        $ret = $this->localcache->hGet($key, $fieldkey);
        $this->assertTrue($ret === $value);
    }

    public function testHashMultiple()
    {
        $this->localcache->select(3);
        $this->localcache->clear();

        $key = 'hashkeymulti';
        $fieldkeyA = 'fieldkeyA';
        $valueA = 'valueA';
        $fieldkeyB = 'fieldkeyB';
        $valueB = 'valueB';
        $fieldkeyC = 'fieldkeyC';

        $ret = $this->localcache->hMSet(
            $key,
            [
                $fieldkeyA => $valueA,
                $fieldkeyB => $valueB,
            ]
        );
        $this->assertTrue($ret !== false);

        $ret = $this->localcache->hMGet(
            $key,
            [
                $fieldkeyA,
                $fieldkeyB,
            ]
        );
        $this->assertTrue(
            !empty($ret)
            && ($ret[$fieldkeyA] === $valueA)
            && ($ret[$fieldkeyB] === $valueB)
        );

        $ret = $this->localcache->hMGet(
            $key,
            [
                $fieldkeyA,
                $fieldkeyB,
                $fieldkeyC,
            ]
        );
        $this->assertTrue(
            !empty($ret)
            && ($ret[$fieldkeyA] === $valueA)
            && ($ret[$fieldkeyB] === $valueB)
            && empty($ret[$fieldkeyC])
        );

        sleep($this->localcache->getLocalCacheTimeout() + 1);

        $ret = $this->localcache->hMGet(
            $key,
            [
                $fieldkeyA,
                $fieldkeyB,
            ]
        );
        $this->assertTrue(
            !empty($ret)
            && ($ret[$fieldkeyA] === $valueA)
            && ($ret[$fieldkeyB] === $valueB)
        );

        $ret = $this->localcache->hMGet(
            $key,
            [
                $fieldkeyA,
                $fieldkeyB,
            ]
        );
        $this->assertTrue(
            !empty($ret)
            && ($ret[$fieldkeyA] === $valueA)
            && ($ret[$fieldkeyB] === $valueB)
        );

        $ret = $this->localcache->hMGet(
            $key,
            [
                $fieldkeyA,
                $fieldkeyB,
                $fieldkeyC,
            ]
        );
        $this->assertTrue(
            !empty($ret)
            && ($ret[$fieldkeyA] === $valueA)
            && ($ret[$fieldkeyB] === $valueB)
            && empty($ret[$fieldkeyC])
        );
    }

    public function testHashGetAll()
    {
        $this->localcache->select(3);
        $this->localcache->clear();

        $key = 'hashkeygetall';
        $fieldkeyA = 'fieldkeyA';
        $valueA = 'valueA';
        $fieldkeyB = 'fieldkeyB';
        $valueB = 'valueB';
        $fieldkeyC = 'fieldkeyC';
        $valueC = 'valueC';

        $ret = $this->localcache->hMSet(
            $key,
            [
                $fieldkeyA => $valueA,
                $fieldkeyB => $valueB,
                $fieldkeyC => $valueC,
            ]
        );
        $this->assertTrue($ret !== false);

        $ret = $this->localcache->hGetAll($key);
        $this->assertTrue(
            !empty($ret)
            && !empty($ret[$fieldkeyA]) && $ret[$fieldkeyA] === $valueA
            && !empty($ret[$fieldkeyB]) && $ret[$fieldkeyB] === $valueB
            && !empty($ret[$fieldkeyC]) && $ret[$fieldkeyC] === $valueC
        );

        $ret = $this->localcache->hGet($key, $fieldkeyA);
        $this->assertTrue($ret === $valueA);
    }

    public function testHashDel()
    {
        $this->localcache->select(3);
        $this->localcache->clear();

        $key = 'hashkeydel';
        $fieldkeyA = 'fieldkeyA';
        $valueA = 'valueA';
        $fieldkeyB = 'fieldkeyB';
        $valueB = 'valueB';
        $fieldkeyC = 'fieldkeyC';
        $valueC = 'valueC';

        $ret = $this->localcache->hMSet(
            $key,
            [
                $fieldkeyA => $valueA,
                $fieldkeyB => $valueB,
                $fieldkeyC => $valueC,
            ]
        );
        $this->assertTrue($ret !== false);

        $ret = $this->localcache->hDel($key, $fieldkeyA, $fieldkeyB);
        $this->assertTrue(!empty($ret));

        $ret = $this->localcache->hGet($key, $fieldkeyA);
        $this->assertTrue(empty($ret));
        $ret = $this->localcache->hGet($key, $fieldkeyB);
        $this->assertTrue(empty($ret));
        $ret = $this->localcache->hGet($key, $fieldkeyC);
        $this->assertTrue($ret === $valueC);
    }

    public function testInvalidRedisHostException()
    {
        $ret = null;

        try {
            $lc = new LocalCache(
                '222.222.222.222',
                'yacPrefixTesting',
                6379,
                3,
                500000,
                3,
                3,
                0
            );
            $lc->select(3);
            $ret = $lc->get('somethingNotExists');
        } catch (\RedisException $e) {
            return $this->assertTrue(true);
        }

        return $this->assertTrue(false);
    }
}


tests/CacheServiceTest.php
<?php
namespace Tests;

use LocalCache\{
    LocalCache,
    CacheService
};
use PHPUnit\Framework\TestCase;

class SampleCache extends CacheService
{
    const PREFIX_FOR_TEST = 'test';

    const CONN_FOR_TEST = 'test_conn';

    const DBIDX_FOR_TEST = 0;

    const EXCEPTION_TRIGGERED = 'here_is_the_exception';

    protected $prefix = self::PREFIX_FOR_TEST;

    protected $connection = self::CONN_FOR_TEST;

    protected $useLocalCache = true;

    protected $db = self::DBIDX_FOR_TEST;

    public $exceptionMsg = '';

    protected function getConfigByConnection(string $connection)
    {
        if ($connection != $this->connection) {
            throw new \Exception('connection name conflicts');
        }

        return [
            'host' => 'localhost',
            'port' => 6379,
            'connectionTimeout' => 1,
            'retryInterval' => 200000,
            'readTimeout' => 1,
            'maxRetry' => 1,
            'reserved' => 0,
        ];
    }

    protected function logException($e)
    {
        assert($e !== false);
        $this->exceptionMsg = self::EXCEPTION_TRIGGERED;
    }
}

// phpcs:ignore
class SampleCacheWithouLocal extends SampleCache
{
    protected $useLocalCache = false;
}

// phpcs:ignore
final class CacheServiceTest extends TestCase
{
    private $cacheWithLocal = null;

    private $cacheWithoutLocal = null;

    private $yac = null;

    private function getYacKey(string $key)
    {
        return SampleCache::DBIDX_FOR_TEST . '_' . SampleCache::PREFIX_FOR_TEST . $key;
    }

    private function getRedisKey($string)
    {
        return SampleCache::PREFIX_FOR_TEST . $string;
    }

    protected function setUp()
    {
        $this->cacheWithLocal = new SampleCache();
        $this->cacheWithoutLocal = new SampleCacheWithouLocal();
        $this->yac = new \Yac(SampleCache::CONN_FOR_TEST);
    }

    public function testSetCacheValueWithLocal()
    {
        $key = 'tscvw';
        $value = 'value_of_tscvw';
        $ttl = 1;
        $defaultValue = 'default_value';

        $ret = $this->cacheWithLocal->setCacheValue($key, $value, $ttl);
        $this->assertTrue($ret !== false);

        $ret = $this->cacheWithLocal->getCacheValue($key);
        $this->assertTrue($ret === $value);

        $ret = $this->cacheWithLocal->get($key);
        $this->assertTrue($ret === $value);

        $ret = $this->yac->get($this->getYacKey($key));
        $this->assertTrue($ret === $value);

        sleep($ttl + 1);
        $ret = $this->cacheWithLocal->getCacheValue($key);
        $this->assertTrue($ret === false);

        $ret = $this->cacheWithLocal->getCacheValue($key, $defaultValue);
        $this->assertTrue($ret === $defaultValue);

        $ret = $this->yac->get($this->getYacKey($key));
        $this->assertTrue(
            $ret === LocalCache::NO_DATA_IN_CACHE || $ret === false
        );
    }

    public function testSetCacheValueWithoutLocal()
    {
        $key = 'tscvw';
        $value = 'value_of_tscvw';
        $ttl = 1;
        $defaultValue = 'default_value';

        $ret = $this->cacheWithoutLocal->setCacheValue($key, $value, $ttl);
        $this->assertTrue($ret !== false);

        $ret = $this->cacheWithoutLocal->getCacheValue($key);
        $this->assertTrue($ret === $value);

        $ret = $this->cacheWithoutLocal->get($key);
        $this->assertTrue($ret === $value);


        $ret = $this->yac->get($this->getYacKey($key));
        $this->assertTrue(
            $ret === LocalCache::NO_DATA_IN_CACHE || $ret === false
        );

        sleep($ttl + 1);
        $ret = $this->cacheWithoutLocal->getCacheValue($key);
        $this->assertTrue($ret === false);

        $ret = $this->cacheWithoutLocal->getCacheValue($key, $defaultValue);
        $this->assertTrue($ret === false);

        $ret = $this->yac->get($this->getYacKey($key));
        $this->assertTrue(
            $ret === LocalCache::NO_DATA_IN_CACHE || $ret === false
        );
    }

    public function testDeleteKeyWithLocal()
    {
        $key = 'tscvw';
        $value = 'value_of_tscvw';

        $this->cacheWithLocal->setCacheValue($key, $value);
        $ret = $this->cacheWithoutLocal->getCacheValue($key);
        $this->assertTrue($ret === $value);

        $this->cacheWithLocal->deleteByKey($key);
        $ret = $this->cacheWithoutLocal->getCacheValue($key);
        $this->assertTrue($ret === false);
        $ret = $this->yac->get($this->getYacKey($key));
        $this->assertTrue($ret === false);
    }

    public function testDeleteKeyWithoutLocal()
    {
        $key = 'tscvw';
        $value = 'value_of_tscvw';

        $this->cacheWithoutLocal->setCacheValue($key, $value);
        $ret = $this->cacheWithoutLocal->getCacheValue($key);
        $this->assertTrue($ret === $value);

        $this->cacheWithoutLocal->deleteByKey($key);
        $ret = $this->cacheWithoutLocal->getCacheValue($key);
        $this->assertTrue($ret === false);
        $ret = $this->yac->get($this->getYacKey($key));
        $this->assertTrue($ret === false);
    }

    public function testException()
    {
        $this->cacheWithLocal->redisCommandDoesNotExists();
        $this->assertTrue(
            $this->cacheWithLocal->exceptionMsg == SampleCache::EXCEPTION_TRIGGERED
        );
    }

    public function testRedisCmdWorking()
    {
        $key = 'tscvw';
        $value = 'value_of_tscvw';

        $this->cacheWithLocal->set($key, $value);
        $ret = $this->cacheWithLocal->get($key);
        $this->assertTrue($ret === $value);
    }
}


src/CachePoolService.php
<?php
namespace LocalCache;

class CachePoolService
{
    /**
     * array of localcache instances
     * array key is redis connection name
     * [
     *      'default' => $LocalCacheInstance
     * ]
     */
    private static $instanceHasCache = [];

    private static $instanceNoCache = [];

    /**
     * get LocalCache Instance
     *
     * @param array $config redis config array. array elements:
     *                      'host' => 'string, MUST return the redis host',
     *                      'port' => 'int, default value is 6379',
     *                      'connectionTimeout' => 'int, default value is 3',
     *                      'retryInterval' => 'int, default value is 500000',
     *                      'readTimeout' => 'int, default value is 3',
     *                      'maxRetry' => 'int, default value is 3',
     *                      'reserved' => 'int, default value is 0',
     * @param bool $useLocalecache set to false will disable locale cache
     * @param string $connection redis connection name
     */
    public static function initCacheInstance(
        array $config,
        bool $useLocalecache = false,
        string $connection = 'default'
    ) {
        if (empty($config['host'])) {
            throw \InvalidArgumentException('unknown host');
        }

        if ($useLocalecache) {
            if (isset(self::$instanceHasCache[$connection])) {
                return self::$instanceHasCache[$connection];
            }

            self::$instanceHasCache[$connection] = new LocalCache(
                $config['host'],
                $connection,
                $config['port'] ?? 6379,
                $config['timeout'] ?? 3,
                $config['retryInterval'] ?? 500000,
                $config['readTimeout'] ?? 3,
                $config['maxRetry'] ?? 3,
                $config['reserved'] ?? 0
            );

            return self::$instanceHasCache[$connection];
        }

        if (isset(self::$instanceNoCache[$connection])) {
            return self::$instanceNoCache[$connection];
        }

        self::$instanceNoCache[$connection] = new LocalCache(
            $config['host'],
            '',
            $config['port'] ?? 6379,
            $config['timeout'] ?? 3,
            $config['retryInterval'] ?? 500000,
            $config['readTimeout'] ?? 3,
            $config['maxRetry'] ?? 3,
            $config['reserved'] ?? 0
        );

        return self::$instanceNoCache[$connection];
    }

    /**
     * get cache instance by connection identifier
     *
     * @param string $connection redis connection identifier
     * @param bool $useLocalecache set true to use local cache
     *
     * @return null|object return LocalCache instance
     */
    public static function getCacheInstance(
        string $connection = 'default',
        bool $useLocalecache = true
    ) {
        if ($useLocalecache) {
            return self::$instanceHasCache[$connection] ?? null;
        }

        return self::$instanceNoCache[$connection] ?? null;
    }

    /**
     * set cache value
     *
     * @param string $connection connection identifier
     * @param int $db database index
     * @param string $key
     * @param mixed $value
     * @param int $ttl
     *
     * @return int|bool
     */
    public static function setCacheValue(
        string $connection,
        int $db,
        string $key,
        $value,
        int $ttl = 0,
        bool $useLocalecache = true
    ) {
        try {
            $rds = self::getCacheInstance($connection, $useLocalecache);
            if (empty($rds)) {
                return false;
            }

            $rds->select($db);
            return $rds->set($key, $value, $ttl);
        } catch (\InvalidArgumentException $e) {
            throw $e;
        }

        return false;
    }

    public static function getCacheValue(
        string $connection,
        int $db,
        string $key,
        $defaultValue = null,
        bool $useLocalecache = true
    ) {
        try {
            $rds = self::getCacheInstance($connection, $useLocalecache);
            if (empty($rds)) {
                return false;
            }

            $rds->select($db);
            return $rds->get($key, $defaultValue);
        } catch (\InvalidArgumentException $e) {
            throw $e;
        }

        return false;
    }

    public static function deleteByKey(
        string $connection,
        int $db,
        string $key,
        bool $useLocalecache = true
    ) {
        try {
            $rds = self::getCacheInstance($connection, $useLocalecache);
            if (empty($rds)) {
                return false;
            }

            $rds->select($db);
            return $rds->delete($key);
        } catch (\InvalidArgumentException $e) {
            throw $e;
        }

        return false;
    }
}


src/CacheService.php
<?php
namespace LocalCache;

abstract class CacheService
{
    /**
     * redis key prefix
     */
    protected $prefix = '';

    /**
     * default database index is 0
     */
    protected $db = 0;

    /**
     * redis configuration name
     */
    protected $connection = 'default';

    /**
     * set false to disable localcache
     */
    protected $useLocalCache = false;

    public function __construct()
    {
        CachePoolService::initCacheInstance(
            $this->getConfigByConnection($this->connection),
            $this->useLocalCache,
            $this->connection
        );
    }

    /**
     * get redis configuration by connection name
     *
     * @param string $connection
     * @return array [
     *      'host' => 'string, MUST return the redis host',
     *      'port' => 'int, default value is 6379',
     *      'connectionTimeout' => 'int, default value is 3',
     *      'retryInterval' => 'int, default value is 500000',
     *      'readTimeout' => 'int, default value is 3',
     *      'maxRetry' => 'int, default value is 3',
     *      'reserved' => 'int, default value is 0',
     * ]
     */
    abstract protected function getConfigByConnection(string $connection);

    /**
     * log redis exception
     *
     * @param InvalidArgumentException|RedisException $e
     * @return none
     */
    abstract protected function logException($e);

    public function setCacheValue(string $key, $value, int $ttl = 0)
    {
        try {
            return CachePoolService::setCacheValue(
                $this->connection,
                $this->db,
                $this->prefix . $key,
                $value,
                $ttl,
                $this->useLocalCache
            );
        } catch (\Exception $e) {
            $this->doLogException($e);
        }

        return false;
    }

    public function getCacheValue(string $key, $defaultValue = null)
    {
        try {
            return CachePoolService::getCacheValue(
                $this->connection,
                $this->db,
                $this->prefix . $key,
                $defaultValue,
                $this->useLocalCache
            );
        } catch (\Exception $e) {
            $this->doLogException($e);
        }

        return false;
    }

    public function deleteByKey(string $key)
    {
        try {
            return CachePoolService::deleteByKey(
                $this->connection,
                $this->db,
                $this->prefix . $key,
                $this->useLocalCache
            );
        } catch (\Exception $e) {
            $this->doLogException($e);
        }

        return false;
    }

    public function __call(string $name, array $arguments)
    {
        $shouldAddPrefix = in_array(
            strtolower($name),
            [
                'hdel',
                'hgetall',
                'hmset',
                'hmget',
                'hset',
                'hget',
                'expire',
                'get',
                'set',
                'delete',
                'del',
                'unlink',
                'yacdeletehash',
            ],
            true
        );
        if ($shouldAddPrefix) {
            $arguments[0] = $this->prefix . $arguments[0];
        }

        $instance = CachePoolService::getCacheInstance(
            $this->connection,
            $this->useLocalCache
        );
        if (empty($instance)) {
            return false;
        }

        try {
            $instance->select($this->db);
            return call_user_func([$instance, $name], ...$arguments);
        } catch (\Exception $e) {
            $this->doLogException($e);
        }

        return false;
    }

    private function doLogException(\Exception $e)
    {
        try {
            $this->logException($e);
        } catch (\Exception $e) {
            // do nothing
            assert(true);
        }
    }
}
